File: pset6/credit.py
import cs50

print("Number: ", end='')
while True:
    num = cs50.get_int()
    if num <= 0:
        print("Retry: ", end='')
    if num > 0:
        break

temp = 1
digits = []
while num > temp:
    digits.append((num // temp) % 10)
    temp *= 10
    
result = [digits[x] if not x % 2 
    else digits[x] * 2 % 10 + digits[x] * 2 // 10  
    for x in range(len(digits)) ]
    
digits.reverse()
if sum(result) % 2 != 0:
    print("INVALID")
elif digits[0] == 4 and (len(digits) == 13 or len(digits) == 16):
    print("VISA")
elif (digits[:2] == [3,7] or digits[:2] == [3,4]) and len(digits) == 15:
    print("AMEX")
elif digits[0] == 5 and (0 < digits[1] < 6) and len(digits) == 16:
    print("MASTERCARD")
else:
    print("INVALID")
        
            

File: pset1/credit/credit.c
# include <stdio.h>

# include <cs50.h>

int main(void)
{
    long long num;
    printf("Number: ");
    do
    {
        num = get_long_long();
    }while(num <= 0);
    
    int count = 0;
    int sum = 0;
    long long i;
    for( i = 1; i <= num; i *= 10)
    {
        int cur_num = (num / i) % 10;
        int flag = count % 2;
        sum = sum + cur_num * ((flag + 1) % 2) 
            + ((cur_num << 1) % 10 + (cur_num << 1) / 10) * flag;
        count++; 
    }
    i /= 100;
    
    if(sum % 10)
    {
        printf("INVALID\n");
        return 0;
    }
    else if(num / (i * 10) == 4 
            && (count == 13 || count == 16))
    {
        printf("VISA\n");
    }
    else if(num / i > 50 && num / i < 56 && count == 16)
    {
        printf("MASTERCARD\n");
    }
    else if((num / i == 34 || num / i == 37) && count == 15)
    {
        printf("AMEX\n");
    }
    else printf("INVALID\n");
    
    
    return 0;
}
